The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5232782788/jobs/9447754434) [(plain)](https://github.com/rust-lang/rust/commit/1e6a863fb9650421caf56fb77618fa451eebd432/checks/14161824429/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling alloc v0.0.0 (/checkout/library/alloc)
error[E0308]: mismatched types
   --> library/std/src/net/tcp/tests.rs:51:38
    |
51  |     match TcpStream::connect_timeout("0.0.0.0:1", Duration::MAX) {
    |           -------------------------- ^^^^^^^^^^^ expected `&SocketAddr`, found `&str`
    |           arguments to this function are incorrect
    |
    = note: expected reference `&core::net::SocketAddr`
               found reference `&'static str`
               found reference `&'static str`
note: associated function defined here
   --> library/std/src/net/tcp.rs:172:12
    |
172 |     pub fn connect_timeout(addr: &SocketAddr, timeout: Duration) -> io::Result<TcpStream> {

For more information about this error, try `rustc --explain E0308`.
error: could not compile `std` (lib test) due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

--- a/library/std/src/net/tcp/tests.rs
+++ b/library/std/src/net/tcp/tests.rs
@@ -47,6 +47,26 @@ fn connect_error() {
 }
 
 #[test]
+fn connect_timeout_error() {
+    let socket_addr = next_test_ip4();
+    let result = TcpStream::connect_timeout(&socket_addr, Duration::MAX);
+    assert!(!matches!(result, Err(e) if e.kind() == ErrorKind::TimedOut));
+
+    let _listener = TcpListener::bind(&socket_addr).unwrap();
+    assert!(TcpStream::connect_timeout(&socket_addr, Duration::MAX).is_ok());
+}
+
+#[test]
+fn connect_timeout_ok_bind() {
+    let listener = TcpListener::bind("127.0.0.1:0").unwrap(); // :0 picks some free port
+    let port = listener.local_addr().unwrap().port(); // obtain the port it picked
+    assert!(
+        TcpStream::connect_timeout(&format!("127.0.0.1:{port}").parse().unwrap(), Duration::MAX)
+            .is_ok()
+    );
+}
+
+#[test]
 fn listen_localhost() {
     let socket_addr = next_test_ip4();
     let listener = t!(TcpListener::bind(&socket_addr));

--- a/library/std/src/sys/windows/net.rs
+++ b/library/std/src/sys/windows/net.rs
@@ -159,7 +159,7 @@ impl Socket {
                 }
 
                 let mut timeout = c::timeval {
-                    tv_sec: timeout.as_secs() as c_long,
+                    tv_sec: cmp::min(timeout.as_secs(), c_long::MAX as u64) as c_long,
                     tv_usec: (timeout.subsec_nanos() / 1000) as c_long,
                 };
 

